### PR TITLE
Update to the new Prometheus library

### PR DESCRIPTION
**Pull request by balbirthomas**
This commit updates kube state metrics to the latest version
of the Prometheus charm library.

### CODE REVIEW

**Comment by balbirthomas** [comment]
Hi @johnsca Here is a tentative set of changes to get Kube State Metrics working again with Current version of Prometheus. I have only sanity tested (screenshot below). Still need to test more thoroughly and fix any unit tests. I just wanted to push this out here to get some feedback. In particular I was not sure the `monitoring_address` property must be used given the new library interface. Hence I have not yet removed that property even though it is unused.
Screenshot from 2021-11-17 19-25-56

**Comment by balbirthomas** [comment]
Fixed unit tests now.

**Comment by addyess** [comment]
@balbirthomas, can you resolve the conflicts in this PR?

**Comment by balbirthomas** [comment]
> @balbirthomas, can you resolve the conflicts in this PR?

Just rebased and pushed. Please do review and merge if you will. Have not yet had the time to run through a full manual end to end test. Though I would expect it should just work.

**Comment by Cynerva** [comment]
Hi @balbirthomas. The integration tests are failing. I've pushed a few fixes for basic maintenance issues, but the tests are still failing somewhere in the new Prometheus library code. Can you look into this?

```
unit-kube-state-metrics-0: 19:08:36 ERROR unit.kube-state-metrics/0.juju-log metrics-endpoint:1: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 114, in <module>
    main(KubeStateMetricsOperator)
  File "/var/lib/juju/agents/unit-kube-state-metrics-0/charm/venv/ops/main.py", line 426, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-kube-state-metrics-0/charm/venv/ops/main.py", line 142, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-kube-state-metrics-0/charm/venv/ops/framework.py", line 276, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-kube-state-metrics-0/charm/venv/ops/framework.py", line 736, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-kube-state-metrics-0/charm/venv/ops/framework.py", line 783, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-kube-state-metrics-0/charm/lib/charms/prometheus_k8s/v0/prometheus_scrape.py", line 1312, in _set_scrape_job_spec
    alert_rules.add_path(self._alert_rules_path, recursive=False)
  File "/var/lib/juju/agents/unit-kube-state-metrics-0/charm/lib/charms/prometheus_k8s/v0/prometheus_scrape.py", line 736, in add_path
    raise InvalidAlertRulePathError(path, "path does not exist")
charms.prometheus_k8s.v0.prometheus_scrape.InvalidAlertRulePathError: path does not exist
```

From the integration test output [here](https://github.com/charmed-kubernetes/kube-state-metrics-operator/runs/4807983091?check_suite_focus=true).

**Review by Cynerva — CHANGES_REQUESTED** [review]
See previous comment about integration test failure.

**Comment by balbirthomas** [comment]
> See previous comment about integration test failure.

Apologies for delay. Bit tied up this week. Will fix and get back to you next.

**Comment by balbirthomas** [comment]
Just updated Prometheus charm library, checked linting and unit tests pass and pushed.

**Comment by balbirthomas** [comment]
@Cynerva I checked that latest branch of Promtheus charm library does indeed trap the `InvalidAlertRulesPath` exception. Not sure how I can trigger the integration tests again on this branch.

**Comment by simonaronsson** [comment]
@Cynerva @balbirthomas:

right now it's failing because the main branch of the actions-operator has been changed from `master` to `main`.

**Comment by addyess** [comment]
https://github.com/charmed-kubernetes/kube-state-metrics-operator/pull/6 opened

**Comment by addyess** [comment]
@balbirthomas can you weigh in on #6.  It's failing the integration tests because of a missing relation?  Seems related to this PR

**Comment by addyess** [comment]
Please rebase and confirm integration tests

**Comment by addyess** [comment]
Well shucks... the integration tests are FURTHER along as now the relations line up -- but there is no resulting data provided from prometheus after 60 seconds of waiting. 

https://github.com/balbirthomas/kube-state-metrics-operator/blob/lma-update/tests/integration/test_charm.py#L39-L55

i see this as the result which indicates that prometheus is answering requests -- just there's nothing in `data.result`
```python
{'status': 'success', 'data': {'resultType': 'vector', 'result': []}}
```

**Comment by balbirthomas** [comment]
> i see this as the result which indicates that prometheus is answering requests -- just there's nothing in `data.result`

Yep. Just ran it locally and it still passes locally. I wonder if this is a timing issue. I am surprised that querying Prometheus should be failing in github but not locally. Locally I am using `microk8s` so that may be one difference. (edit) Just noticed github workflow is using microk8s too.

**Comment by balbirthomas** [comment]
The default `scrape_interval` is 1 minute. This could be the issue since all 12 Prometheus queries (with 5 second sleeps) will finish by the end of the first minute. Should I reduce the `scrape_interval` to 5 seconds or would you rather extend the sleep time to some thing like 30 seconds ?

**Comment by addyess** [comment]
Lets have the integration tests configure the scrape_interval to be 5 in order to match the test:

```python
async def test_stats_in_prometheus(ops_test, helpers):
    result = None
    for attempt in range(12):
        ...
        await asyncio.sleep(5)  # the scrape interval is 5s          <<< -----------
    else:
        pytest.fail(f"Failed to get expected query result: {result}")

**Comment by addyess** [comment]
Maybe this??

````python 
    await ops_test.model.deploy(
        prometheus_charm,
        resources={
            "prometheus-image": "prom/prometheus",
        },
        config={"scrape-interval": "5s"},
    )
```

**Comment by addyess** [comment]
Seems like that scrape metric should be configurable by this charm and provided over the Metric endpoint?

File: src/charm.py [hunk 1]
```diff
@@ -29,19 +29,23 @@ class KubeStateMetricsOperator(CharmBase):
...
-        )
+        jobs = [
+            {
+                "static_configs": [
+                    {
+                        "targets": ["*:8080", "*:8081"],
+                    }
+                ]
+            }
+        ]
```
Comment (addyess) [inline src/charm.py:40]:
```suggestion
        jobs = [
            {
                "scrape_interval": self.model.config["scrape-interval"],
                "static_configs": [
                    {
                        "targets": ["*:8080", "*:8081"],
                    }
                ]
            }
        ]
```

**Review by addyess — CHANGES_REQUESTED** [review]
seems we need a `scrape-interval` in config.yaml

```yaml
options:
  scrape-interval:
    default: 1m
    description: "Prometheus job configuration for scrape interval"
    type: string
```

**Comment by balbirthomas** [comment]
@addyess We can indeed add `scrape-interval` to the kube-state-metrics `config.yaml` and use it. I am going to add it to kube-state-metrics run locally and then make a push. Since I have been unable to reproduce the issue locally this is the only way I can test the hypothesis.

**Comment by addyess** [comment]
@balbirthomas i've been making a change to a PR in your branch.  see https://github.com/balbirthomas/kube-state-metrics-operator/pull/2

**Comment by addyess** [comment]
closed in favor of https://github.com/charmed-kubernetes/kube-state-metrics-operator/pull/7

**Pull request closed without merging**